### PR TITLE
DDP-related improvements to data module and logging

### PR DESCRIPTION
**Pull request by tesfaldet**
## What does this PR do?

1. Changed the pylogger to be rank-aware, such that logged messages will have the rank of the process prefixed. It also has the capability of logging to a specific rank if the user wishes to, thereby covering the previous pylogger's use-case of rank-zero logging while providing greater logging flexibility. By default, the new pylogger is unrestricted in the ranks it logs to so as to provide greater clarity as to which process the current log is being executed in (e.g., when instantiating models and such, it's useful to know that it's happening on multiple processes).
2. The MNIST DataModule's batch size is now divided by the number of processes used in a DDP setup, so as to keep training dynamics more consistent/comparable when running multiple training scripts, each with a different number of devices.
3. The log file saved by the Hydra colorlog plugin is now consistent across devices when training in a DDP setup. This means that all processes will log to the same file, and because the new pylogger provides rank information, it is easy for the user to tell from which process the log came from. This is in contrast to before where one file called `train.log` would be created for the rank 0 process and `train_ddp_process_{rank}.log` would be created for all the other ranks, making it confusing to read through logs in a DDP setup.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [~] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?
y

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tesfaldet** [comment]
Also I just realized I might need to take this issue into account https://github.com/Lightning-AI/lightning/issues/12862. This affects the train script because `trainer.test` is called right after `trainer.fit` is finished. If the trainer was initialized with the DDP strategy, then when `test` starts it will give a warning:

```
lightning/pytorch/trainer/connectors/data_connector.py:226: PossibleUserWarning: Using `DistributedSampler` with the dataloaders. During `trainer.test()`, it is recommended to use `Trainer(devices=1, num_nodes=1)` to ensure each sample/batch gets evaluated exactly once. Otherwise, multi-device settings use `DistributedSampler` that replicates some samples to make sure all devices have same batch size in case of uneven inputs.
```

I think this warning is quite important to deal with since we wouldn't want inaccurate test metrics to be reported because of accidentally re-using the same trainer that was initialized to use a DDP strategy. A possible solution is to initialize a new trainer separately for testing. The only issue I can think of with this approach is that it will create a new Logger for testing, meaning that you won't have all your train, validation, and test results neatly presented in a single log (e.g., a single TensorBoard log, or a single WandB log, or a single Aim log). Lightning doesn't save Logger objects in checkpoints, meaning they can't be restored from checkpoints (although I'm aware of [your PR for this](https://github.com/Lightning-AI/lightning/pull/13069), but it's specific to the Tensorboard Logger it seems).

File: src/utils/pylogger.py [hunk 1]
```diff
@@ -1,21 +1,50 @@
...
     """
+    T = TypeVar("T")
+    P = ParamSpec("P")
+
+    def _rank_prefixed_log(fn: Callable[P, T]) -> Callable[P, Optional[T]]:
+        """Wrap a logging function to prefix its message with the rank of the process it's being
+        logged from.
+
+        If `'rank'` is provided in the wrapped functions kwargs, then the log will only occur on
+        that rank/process.
```
Comment (ashleve) [inline src/utils/pylogger.py:24]: This confuses me - what's the default way for user to provide 'rank'?

**Review by ashleve — CHANGES_REQUESTED** [review]
1. I like this new logger. What's the default way for user to limit the logging to only master rank? Can you add something like optional `log_master_only` arg to `get_ranked_pylogger(...)`?
2. Sure, we can have this
3. Good idea. Are you sure logging to the same file from many processes doesn't need synchronisation? Won't there by any conflicts leading to logs getting lost sometimes?

**Comment by ashleve** [comment]
image

**Comment by tesfaldet** [comment]
Getting to it now! Sorry for the hold up.

**Comment by tesfaldet** [comment]
> 1. I like this new logger. What's the default way for user to limit the logging to only master rank? Can you add something like optional `log_master_only` arg to `get_ranked_pylogger(...)`?
> 2. Sure, we can have this
> 3. Good idea. Are you sure logging to the same file from many processes doesn't need synchronisation? Won't there by any conflicts leading to logs getting lost sometimes?

1. Thanks! I can certainly provide an optional argument to `get_ranked_pylogger(...)` to restrict ranking to rank zero, so that the user doesn't always have to keep providing `rank=0` to the log function. How does `rank_zero_only` sound? That should keep things in-line with naming convention across Lightning so the user immediately knows what the optional arg means.

2. 👍 

3. So far in my hundreds of DDP experiments I haven't seen any issues regarding file writes to the same log file from multiple process. Even from hundreds of processes at a time. However, this is just anecdotal. Let me see if I can find code that shows us that there's some sync or atomized file-writing going on under the hood. I'm guessing I'm going to have to look at `logging.FileHandler` as that's what it looks like Hydra is using. I'll report back.

**Comment by ashleve** [comment]
yea `rank_zero_only` is better

**Comment by tesfaldet** [comment]
Added `rank_zero_only` :) I also changed the implementation since I was having issues with the previous implementation I had.

**Review by ashleve — APPROVED** [review]
LGTM

**Comment by tesfaldet** [comment]
I forgot to report back about the multiprocessing logging concern. In short, for the context of distributed logging where it's ok to have interleaving logs from multiple processes (i.e., you're not expecting a guaranteed ordering of logs across processes), then all seems to be ok. Check this stack overflow post for more info https://stackoverflow.com/questions/12238848/python-multiprocessinglogging-filehandler. Furthermore, there was some discussions surrounding Hydra configuring its logging setup to support logging in a distributed setting, but it didn't result in any concrete change https://github.com/facebookresearch/hydra/issues/1148. From what I and others have experienced, nothing bad has happened to our logs when logging from multiple processes to a single file.

That being said, it might be worth looking into the officially-recommended way of logging to a single file from multiple processes: https://docs.python.org/3/howto/logging-cookbook.html#logging-to-a-single-file-from-multiple-processes
(that shouldn't be the responsibility of this template though...)

**Pull request merged**